### PR TITLE
New rule: use_function_declaration_instead_of_variable

### PR DESCRIPTION
**Pull request by JPaulsen**
Fixes https://github.com/dart-lang/linter/issues/459

### CODE REVIEW

File: lib/src/rules/use_function_declaration_instead_variable.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    if (node.rightHandSide.bestType is FunctionType) {
```
Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: @bwilkerson @alexeieleusis do you know if there is a better way to know if something is a function? 
'''
implementsInterface(node.rightHandSide.bestType, 'Function', 'dart.core')
'''
did not work, and
'''
node.rightHandSide.bestType.isDartFunction
'''
Neither

Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: @bwilkerson @alexeieleusis do you know a better way to know if a type is a function?

Both of these didn't work
```
DartTypeUtilities.implementsInterface(node.rightHandeSide.bestType, 'Function', 'dart.core');
node.rightHandeSide.bestType.isDartFunction;
```

Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Also, in a FunctionType, name = null and element.name = ''

File: lib/src/rules/use_function_declaration_instead_variable.dart [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
```
Comment (bwilkerson) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Don't produce a lint for assignments. Code to assign a function to a variable can't always be replaced by a function declaration. Consider
```
class Dispatcher {
  Map<String, Function> callbacks = {};

  void on(String event, Function callback) {
    callbacks[event] = callback; // OK
  }
}
```

File: lib/src/rules/use_function_declaration_instead_variable.dart [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    if (node.rightHandSide.bestType is FunctionType) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  visitVariableDeclaration(VariableDeclaration node) {
+    if (node.initializer != null && node.initializer.bestType is FunctionType) {
```
Comment (bwilkerson) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: I don't think you want to test the type of the initializer here at all. I think you want to ask `node.initializer is FunctionExpression`.

Otherwise you'll end up creating a lint for something like
```
void doIt(Function f) {
  var x = f; // OK
  x(this);
}
```

File: lib/src/rules/use_function_declaration_instead_variable.dart (hunk 1 above)
Comment (bwilkerson) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: You don't actually need to do either, but...

In analyzer, just like in Dart, the types are divided into `InterfaceType`s and `FunctionType`s. If a type is a `FunctionType`, then it's a function, but it's also a function if it's an `InterfaceType` that implements the interface `Function` or defined a method named `call`. The first code snippet does part of that work for you. The second, however, tests to see whether the type is exactly the `InterfaceType` representing the class `Function` (and will return false for subtypes of `Function`).

File: lib/src/rules/use_function_declaration_instead_variable.dart (hunk 3 above)
Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Yes, I intentionally created them, can you please comment the test to know exactly what is right and wrong?

File: test/rules/use_function_declaration_instead_variable.dart [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N use_function_declaration_instead_variable`
+
+void main() {
+  thisIsGood() { // OK
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: Agreed that this is OK.

File: test/rules/use_function_declaration_instead_variable.dart [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    print('this is good');
+  }
+
+  var bad1 = () { // LINT
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: Agreed that this should produce a lint.

File: test/rules/use_function_declaration_instead_variable.dart [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    print('this is bad');
+  };
+
+  var bad2 = bad1; // LINT
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: I don't think this should be a lint. The only "fix" for this is
```
void bad2() {
  bad1();
}
```
which I don't believe the guideline is advocating.

File: test/rules/use_function_declaration_instead_variable.dart [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+  bad2 = bad1; // LINT
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: This shouldn't be a lint because there's nothing else that the user can write. You can't have
```
var x = ...;
int x(String s) { ... }
```
because then you'd have two definitions of the same name (the local variable declaration and the local function declaration). If the value of `x` needs to change at this point, an assignment is the only option.

File: test/rules/use_function_declaration_instead_variable.dart [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+  bad2 = () { // LINT
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: This shouldn't be a lint for the same reason. You can't create lints for assignments because there is no alternative with the same semantics.

File: test/rules/use_function_declaration_instead_variable.dart [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    print('this is bad');
+  };
+
+  var bad3 = print; // LINT
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: Again, I don't think preventing this is the intent of the guideline. Consider
```
var write = print;
if (args['writeToFile']) {
  write = (String s) {
    // write `s` to a file
  }
}
doWork(write);
```
If we define a local function named `write`, then we cannot re-assign it before passing it in somewhere. (Which means, by the way, that the lint should also not be produced when the variable is assigned another value even if the initializer contains a function literal.)

The point of the guideline is that if you're going to define a function, it's easier to read it if it looks like a function definition. The point isn't to prevent users from defining and using function-valued variables.

File: test/rules/use_function_declaration_instead_variable.dart [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  bad3 = print; // LINT
```
Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: Again, assignments cannot produce lints.

File: test/rules/use_function_declaration_instead_variable.dart (hunk 8 above)
Comment (JPaulsen) [inline test/rules/use_function_declaration_instead_variable.dart]: Are you sure about line 20?, for example, I can do this:

```
var bad2;
bad2 = () {
 print('this is bad');
}
```

but not this:
```
var bad1 = () {
 print('this is bad');
}
```

Comment (bwilkerson) [inline test/rules/use_function_declaration_instead_variable.dart]: I believe that's reasonable (that the first does not produce a lint while the second does). Again, the point of the guideline (as I understand it) is *not* to prevent assigning a function to a variable, the point is to encourage code that is easier to read. The hypothesis, which I agree with, is that this
```
void bad1() {
  print('this is good');
}
```
is easier to read than
```
var bad1 = () {
 print('this is bad');
}
```
But I don't think we should generate a lint if the variable is re-assigned somewhere later in the function / method, because then the user really *is* trying to create a variable, not just a function.

To put it another way, the guideline is about understanding user intent and helping them accomplish their goal in a better way. In order to do that we have to understand whether they're trying to define a function or a variable.

File: lib/src/rules/use_function_declaration_instead_variable.dart (hunk 1 above)
Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Fixed

File: lib/src/rules/use_function_declaration_instead_variable.dart (hunk 2 above)
Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Fixed

File: lib/src/rules/use_function_declaration_instead_variable.dart (hunk 3 above)
Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Fixed

File: test/rules/use_function_declaration_instead_variable.dart (hunk 6 above)
Comment (JPaulsen) [inline test/rules/use_function_declaration_instead_variable.dart]: Fixed

File: test/rules/use_function_declaration_instead_variable.dart (hunk 7 above)
Comment (JPaulsen) [inline test/rules/use_function_declaration_instead_variable.dart]: Fixed

File: test/rules/use_function_declaration_instead_variable.dart (hunk 8 above)
Comment (JPaulsen) [inline test/rules/use_function_declaration_instead_variable.dart]: Fixed

File: test/rules/use_function_declaration_instead_variable.dart (hunk 9 above)
Comment (JPaulsen) [inline test/rules/use_function_declaration_instead_variable.dart]: Fixed

File: test/rules/use_function_declaration_instead_variable.dart (hunk 10 above)
Comment (JPaulsen) [inline test/rules/use_function_declaration_instead_variable.dart]: Fixed

File: lib/src/rules/use_function_declaration_instead_variable.dart [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+}
+```
+
+''';
+
+class UseFunctionDeclarationInsteadVariable extends LintRule {
+  _Visitor _visitor;
+  UseFunctionDeclarationInsteadVariable()
+      : super(
+            name: 'use_function_declaration_instead_variable',
```
Comment (bwilkerson) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Perhaps "use_function_declaration_instead_of_variable".

Comment (JPaulsen) [inline lib/src/rules/use_function_declaration_instead_variable.dart]: Fixed
